### PR TITLE
docs(testworkflows): add more documentation about TestWorkflows

### PR DESCRIPTION
**Pull request by rangoo94**
## Pull request description 

* Add more documentation about TestWorkflow expressions

## Checklist (choose whats happened)

- [ ] breaking change! (describe)
- [ ] tested locally
- [ ] tested on cluster
- [ ] added new dependencies
- [x] updated the docs
- [ ] added a test

### CODE REVIEW

**Comment by rangoo94** [comment]
Hi @jfermi, could you check? Thanks in advance! 🙂

File: docs/docs/articles/test-workflows-examples-expressions.md [hunk 1]
```diff
@@ -14,40 +14,158 @@ You can do basic math easily, like **config.workers * 5**.
 
 ![Expressions](../img/expressions.png) 
 
+### Operators
+
+#### Arithmetic
+
+The operators have the precedence defined, so the order will follow math rules, as an example:
```
Comment (jfermi) [inline docs/docs/articles/test-workflows-examples-expressions.md:21]:
```suggestion
The operators have the precedence defined so the order will follow math rules. Examples:
```

File: docs/docs/articles/test-workflows-examples-expressions.md [hunk 2]
```diff
@@ -14,40 +14,158 @@ You can do basic math easily, like **config.workers * 5**.
...
+| `services` (like `services.db.0.ip` or `services.db.*.ip`) | ❌                    | Get the IPs of initialized services                                               |
 
 ### Contextual Variables
 
 In some contexts, there are additional variables available.
 
-As an example, while writing the condition, you can use variables like passed (bool), failed (bool), always (true), never (false), status (string) that refer to current status of the TestWorkflow.
-
-![Built-in Variables](../img/built-in-variables.png) 
+#### Retry conditions
```
Comment (jfermi) [inline docs/docs/articles/test-workflows-examples-expressions.md:83]:
```suggestion
#### Retry Conditions
```

File: docs/docs/articles/test-workflows-examples-expressions.md [hunk 3]
```diff
@@ -14,40 +14,158 @@ You can do basic math easily, like **config.workers * 5**.
...
+spec:
+  steps:
+  - shell: exit 0
+    # ensure that the step won't fail for 5 executions
+    retry:
+      count: 5
+      until: 'self.failed' 
+```
+
+#### Matrix and shards
```
Comment (jfermi) [inline docs/docs/articles/test-workflows-examples-expressions.md:97]:
```suggestion
#### Matrix and Shard
```

File: docs/docs/articles/test-workflows-examples-expressions.md [hunk 4]
```diff
@@ -14,40 +14,158 @@ You can do basic math easily, like **config.workers * 5**.
...
+
+When using `services` (service pods), `parallel` (parallel workers), or `execute` (test suite) steps:
+
+* You can use `matrix.<name>` and `shard.<name>` to access parameters for each copy
```
Comment (jfermi) [inline docs/docs/articles/test-workflows-examples-expressions.md:101]:
```suggestion
* You can use `matrix.<name>` and `shard.<name>` to access parameters for each copy.
```

File: docs/docs/articles/test-workflows-examples-expressions.md [hunk 5]
```diff
@@ -14,40 +14,158 @@ You can do basic math easily, like **config.workers * 5**.
...
+* You can access `index` and `count` that will differ for each copy
```
Comment (jfermi) [inline docs/docs/articles/test-workflows-examples-expressions.md:102]:
```suggestion
* You can access `index` and `count` that will differ for each copy.
```

File: docs/docs/articles/test-workflows-examples-expressions.md [hunk 6]
```diff
@@ -14,40 +14,158 @@ You can do basic math easily, like **config.workers * 5**.
...
+* Also, you may use `matrixIndex`, `matrixCount`, `shardIndex` and `shardCount` to get specific indexes/numbers for combinations and shards
```
Comment (jfermi) [inline docs/docs/articles/test-workflows-examples-expressions.md:103]:
```suggestion
* Also, you may use `matrixIndex`, `matrixCount`, `shardIndex` and `shardCount` to get specific indexes/numbers for combinations and shards. 
```

**Review by jfermi — COMMENTED** [review]
A few editing suggestions.

**Review by vsukhin — APPROVED** [review]

**Pull request merged**